torchgeo/datasets/patternnet.py

```diff
@@ -0,0 +1,174 @@
+    2. basketball_court
+    3. beach
+    4. bridge
+    5. cemetery
+    6. chaparral
+    7. christmas_tree_farm
+    8. closed_road
+    9. coastal_mansion
+    10. crosswalk
+    11. dense_residential
+    12. ferry_terminal
+    13. football_field
+    14. forest
+    15. freeway
+    16. golf_course
+    17. harbor
+    18. intersection
+    19. mobile_home_park
+    20. nursing_home
+    21. oil_gas_field
+    22. oil_well
+    23. overpass
+    24. parking_lot
+    25. parking_space
+    26. railway
+    27. river
+    28. runway
+    29. runway_marking
+    30. shipping_yard
+    31. solar_panel
+    32. sparse_residential
+    33. storage_tank
+    34. swimming_pool
+    35. tennis_court
+    36. transformer_station
+    37. wastewater_treatment_plant
+
+    If you use this dataset in your research, please cite the following paper:
+    * https://www.sciencedirect.com/science/article/abs/pii/S0924271618300042
+    """
+
+    url = "https://drive.google.com/file/d/127lxXYqzO6Bd0yZhvEbgIfz95HaEnr9K"
+    md5 = "96d54b3224c5350a98d55d5a7e6984ad"
+    filename = "PatternNet.zip"
+    files = ["images"]
+
+    def __init__(
+        self,
+        root: str = "data",
+        transforms: Optional[Callable[[Tensor], Tensor]] = None,
+        download: bool = False,
+        checksum: bool = False,
+    ) -> None:
+        """Initialize a new PatternNet dataset instance.
+
+        Args:
+            root: root directory where dataset can be found
+            transforms: a function/transform that takes input sample and its target as
+                entry and returns a transformed version
+            download: if True, download dataset and store it in the root directory
+            checksum: if True, check the MD5 of the downloaded files (may be slow)
+
+        Raises:
+            RuntimeError: if ``download=False`` and data is not found, or checksums
+                don't match
+        """
+        self.root = root
+        self.checksum = checksum
+
+        if download:
+            self._download()
+
+        if not self._check_integrity():
+            raise RuntimeError(
+                "Dataset not found or corrupted. "
+                + "You can use download=True to download it"
+            )
+
+        super().__init__(
+            root=os.path.join(root, "images"),
+            transform=PILToTensor(),
+            target_transform=torch.tensor,  # type: ignore[attr-defined]
+        )
+
+        # Must be set after calling super().__init__()
+        self.transforms = transforms
+
+    def __getitem__(self, index: int) -> Dict[str, Tensor]:
+        """Return an index within the dataset.
+
+        Args:
+            index: index to return
+
+        Returns:
+            data and label at that index
+        """
+        image, label = ImageFolder.__getitem__(self, index)
+
+        if self.transforms is not None:
+            image = self.transforms(image)
+
+        sample = {"image": image, "label": label}
```

Comment:
`self.transforms` is designed to take in a sample dict as input, not a single image.

Reply:
Do we want it to act this way for image classification datasets? I understand why it would be this way for detection/segmentation. This seems like it would force the user to roll their own transforms instead of just using transforms = T.Compose([ToTensor(), T.Normalize()])

Reply:
We want the transforms provided with TorchGeo to work for all datasets, regardless of what type of task we are trying to solve. See https://discuss.pytorch.org/t/state-of-the-art-for-torchvision-datasets-transforms-models-design/123625 for some discussion on this. At the moment, our transform support is pretty poor, but I think my next task will be wrapping the transforms that come with Kornia so that users can use them directly. I'll also be adding geospatial-specific transforms.

Reply:
Ok that's fine, I can change it back to use a dict input.

Kornia is awesome. I've used it a ton. But any of their transforms should probably be used after a batch is loaded, no? So that you can use the GPU to perform the normalization and augmentations.

Reply:
That's another thing we need to think about. Right now, ZipDataset is designed in such a way that you pass the transforms to each individual GeoDataset separately, which means that things like RandomHorizontalFlip won't work correctly for Landsat + CDL. One possible solution is to compute the transform after the batch is loaded, but then you can't benefit from data loader parallelization. I haven't yet used Kornia myself so I still need to look into this.

Reply:
In my experience it's actually more efficient to take advantage of the multiprocessing in the dataloader only for loading images, and then utilize kornia on the GPU to perform all normalizations & augmentations on a batched tensor. Particularly in the self-supervised learning area where it's common to chain multiple augmentations that significantly slow down CPU data loading (to grayscale + random flips + random resized crop + color jitter). 

The other nice features of kornia which I have been using recently are:
- [kornia.augmentation.VideoSequential](https://kornia.readthedocs.io/en/latest/augmentation.container.html#kornia.augmentation.container.VideoSequential): repeat or vary augmentation per image in time series/video
- [kornia.augmentation.PatchSequential](https://kornia.readthedocs.io/en/latest/augmentation.container.html#kornia.augmentation.container.PatchSequential): repeat or vary augmentation per patch in image
- [kornia.contrib.ExtractTensorPatches](https://kornia.readthedocs.io/en/latest/contrib.html#kornia.contrib.ExtractTensorPatches): extract patches from each image in batch, useful for vision datasets where the images are very large (e.g. 2000x2000) but you want to train on patches (e.g. 256x256)

If using a PyTorch Lightning DataModule for training you can add the augmentations as an input arg and easily use them in the [LightningDataModule.on_after_batch_transfer](https://pytorch-lightning.readthedocs.io/en/latest/extensions/datamodules.html#on-after-batch-transfer) method.

Reply:
Updated to pass sample dict of image and label to transforms. I read the link you posted on the torch forums and I agree this is the way we should standardize.